pennylane/tape/functions/unified.py

```diff
@@ -0,0 +1,210 @@
+            unwrapped.
+
+        **Example**
+
+        >>> x = tf.Variable([0.4, 0.1, 0.5])
+        >>> y = UnifiedTensor(x)
+        >>> z = tf.constant([0.1, 0.2])
+
+        Note that this is a static method, so we must pass the unified tensors itself
+        if we would like it to be included.
+
+        >>> res = y.unwrap([y, z])
+        >>> res
+        [<tf.Variable 'Variable:0' shape=(3,) dtype=float32, numpy=array([0.4, 0.1, 0.5], dtype=float32)>,
+         <tf.Tensor: shape=(2,), dtype=float32, numpy=array([0.1, 0.2], dtype=float32)>]
+        >>> print([type(v) for v in res])
+        [<class 'tensorflow.python.ops.resource_variable_ops.ResourceVariable'>,
+         <class 'tensorflow.python.framework.ops.EagerTensor'>]
+        """
+        return [v.data if isinstance(v, UnifiedTensor) else v for v in tensors]
+
+    @property
+    @abc.abstractmethod
+    def shape(self):
+        """Returns the shape of the tensor as a tuple of integers."""
+
+    @property
+    @abc.abstractmethod
+    def T(self):
+        """Returns the transpose of the tensor."""
+
+    @abc.abstractmethod
+    def expand_dims(self, axis):
+        """Expand the shape of the tensor.
+
+        Args:
+            axis (int or tuple[int]): the axis or axes where the additional
+                dimensions should be inserted
+        """
+
+    @abc.abstractmethod
+    def numpy(self):
+        """Converts the tensor to a standard, non-differentiable NumPy ndarray or Python scalar if
+        the tensor is 0-dimensional.
+
+        Returns:
+            array:
+
+        **Example**
+
+        >>> x = tf.Variable([0.4, 0.1, 0.5])
+        >>> y = UnifiedTensor(x)
+        >>> y.numpy()
+        array([0.4, 0.1, 0.5], dtype=float32)
+        """
+
+    @abc.abstractmethod
+    def ones_like(self):
+        """Returns a unified tensor of all ones, with the shape and dtype
+        of the unified tensor.
+
+        Returns:
+            UnifiedTensor: all ones array
+
+        **Example**
+
+        >>> x = tf.Variable([[0.4, 0.1], [0.1, 0.5]])
+        >>> y = UnifiedTensor(x)
+        >>> y.ones_like()
+        tf.Tensor(
+        [[1. 1.]
+         [1. 1.]], shape=(2, 2), dtype=float32)
+        """
+
+    @staticmethod
+    @abc.abstractmethod
+    def stack(values, axis=0):
+        """Stacks a list of tensors along the specified index.
+
+        Args:
+            values (Sequence[array_like]): sequence of arrays/tensors to stack
+            axis (int): axis on which to stack
+
+        **Example**
+
+        >>> x = tf.Variable([0.4, 0.1, 0.5])
+        >>> a = tf.constant([1., 2., 3.])
+        >>> y = UnifiedTensor(x)
+        >>> y.stack([a, y])
+        <tf.Tensor: shape=(2, 3), dtype=float32, numpy=
+        array([[1. , 2. , 3. ],
+               [0.4, 0.1, 0.5]], dtype=float32)>
+
+        Note that this is a static method, so we must pass the unified tensor itself
+        if we would like it to be included.
+        """

```

Comment:
That's a very cool idea to make these abstract methods!

Reply:
It allows us to 

* enforce the same signature, 
* have a source of truth for how it should behave, and 
* have a common docstring!